File: leetcode/range-sum-query-immutable.py
# 303. Range Sum Query - Immutable
# 🟢 Easy
#
# https://leetcode.com/problems/range-sum-query-immutable/
#
# Tags: Array - Design - Prefix Sum

import timeit
from itertools import accumulate
from typing import List


# Create an array of sums. When a range sum is requested, fetch it
# from Sum(right - left).
#
# Time complexity: O(n) - To create the prefix sum array, O(1) to
# calculate the range sums on sumRange() calls.
# Space complexity: O(n) for the prefix sum array.
#
# Runtime 76 ms Beats 90.39%
# Memory 17.7 MB Beats 59.83%
class NumArray:
    def __init__(self, nums: List[int]):
        self.sums = list(accumulate(nums))

    def sumRange(self, left: int, right: int) -> int:
        return self.sums[right] - (self.sums[left - 1] if left > 0 else 0)


# Similar idea but use a zero padded sums array instead of checking if
# the value is positive when computing the sum.
#
# Runtime 72 ms Beats 96.72%
# Memory Usage 17.7 MB Beats 25.94%
class ZeroPadded:
    def __init__(self, nums: List[int]):
        self.sums = list(accumulate([0] + nums))

    def sumRange(self, left: int, right: int) -> int:
        return self.sums[right + 1] - self.sums[left]


def test():
    executors = [
        NumArray,
        ZeroPadded,
    ]
    tests = [
        [
            [2],
            {
                2: [0, 0],
            },
        ],
        [
            [-2, 0, 3, -5, 2, -1],
            {
                1: [0, 2],
                -1: [2, 5],
                -3: [0, 5],
                -1: [5, 5],
            },
        ],
    ]
    for executor in executors:
        start = timeit.default_timer()
        for _ in range(1):
            for col, t in enumerate(tests):
                sol = executor(t[0])
                for sum in t[1]:
                    result = sol.sumRange(*t[1][sum])
                    exp = sum
                assert result == exp, (
                    f"\033[93m» {result} <> {exp}\033[91m for"
                    + f" test {col} using \033[1m{executor.__name__}"
                )
        stop = timeit.default_timer()
        used = str(round(stop - start, 5))
        cols = "{0:20}{1:10}{2:10}"
        res = cols.format(executor.__name__, used, "seconds")
        print(f"\033[92m» {res}\033[0m")


test()

File: leetcode/range-sum-query-immutable.rs
// 303. Range Sum Query - Immutable
// 🟢 Easy
//
// https://leetcode.com/problems/range-sum-query-immutable/
//
// Tags: Array - Design - Prefix Sum

struct NumArray {
    prefix_sums: Vec<i32>,
}

/**
 * Create an array of sums. When a range sum is requested, fetch it from
 * prefix_sums(right - left).
 *
 * Time complexity: O(n) - To create the prefix sum array we iterate over
 * the entire input array, then O(1) for the sum_range method.
 * Space complexity: O(n) - We store an array of prefix sums of the same
 * size as the input array.
 *
 * Runtime 6 ms Beats 77.78%
 * Memory 8.4 MB Beats 90.48%
 */
impl NumArray {
    /**
     * The constructor creates a vector of prefix sums.
     */
    fn new(nums: Vec<i32>) -> Self {
        NumArray {
            prefix_sums: nums
                .iter()
                .scan(0, |state, &x| {
                    *state = *state + x;
                    Some(*state)
                })
                .collect(),
        }
    }

    /**
     * Use the prefix sums to compute the result in O(1)
     */
    fn sum_range(&self, left: i32, right: i32) -> i32 {
        self.prefix_sums[right as usize]
            - if left > 0 {
                self.prefix_sums[left as usize - 1]
            } else {
                0
            }
    }
}

// Tests.
fn main() {
    let num_array = NumArray::new(vec![-2, 0, 3, -5, 2, -1]);
    assert_eq!(num_array.sum_range(0, 2), 1);
    assert_eq!(num_array.sum_range(2, 5), -1);
    assert_eq!(num_array.sum_range(0, 5), -3);
    println!("All tests passed!")
}
